### PR TITLE
Adding Multichip Microbenchmarks

### PR DESCRIPTION
**Pull request by SeanNijjar**
- Bidirectional bandwidth microbenchmarks
   - This benchmark replays buffers from local erisc over the ethernet link. No data integrity check (since we don't have overheads of receiving data from noc)
   - Peak acheived was ~21.3GBps for best configuration
   - No IRAM support enabled yet

```
TT_METAL_DEVICE_PROFILER=1 {os.environ['TT_METAL_HOME']}/build/test/tt_metal/perf_microbenchmark/ethernet/test_ethernet_bidirectional_bandwidth_no_edm {num_samples}  {sample_size}  {num_channels}
```  

Latency benchmark:
```
TT_METAL_DEVICE_PROFILER=1  ./build/test/tt_metal/perf_microbenchmark/test_ethernet_hop_latencies_no_edm  {len(sample_counts)} {sample_counts_str}    {len(page_sizes)} {page_sizes_str}   {len(channel_counts)} 
 {channel_counts_str}   {len(hop_counts)} {hop_counts_str}
```

Initial Latency Results:
https://docs.google.com/spreadsheets/d/16gRDxhMGRUYQbLZeGcjvmWUY5onnmWM-b3Zfa4jJCX8/edit?usp=sharing
image

image

image

image

### CODE REVIEW

**Comment by SeanNijjar** [comment]
workflows:
post-commit: https://github.com/tenstorrent/tt-metal/tree/snijjar/ccl-microbenchmarks
multi-device unit tests: https://github.com/tenstorrent/tt-metal/actions/runs/8714494939
microbench: https://github.com/tenstorrent/tt-metal/actions/runs/8714480105

File: tests/tt_eager/module.mk [hunk 1]
```diff
@@ -29,7 +29,7 @@ TT_EAGER_TESTS += \
 TT_EAGER_TESTS_SRCS = $(addprefix tests/tt_eager/, $(addsuffix .cpp, $(TT_EAGER_TESTS:tests/%=%)))
 
 TT_EAGER_TESTS_INCLUDES = $(TEST_INCLUDES) $(TT_EAGER_INCLUDES)
-TT_EAGER_TESTS_LDFLAGS = $(TT_METAL_TESTS_LDFLAGS) $(TT_LIB_LDFLAGS) -lgtest -lgtest_main
+TT_EAGER_TESTS_LDFLAGS = $(TT_METAL_TESTS_LDFLAGS) $(TT_LIB_LDFLAGS) -lgtest -lgtest_main -ltt_lib_csrc
```
Comment (tt-rkim) [inline tests/tt_eager/module.mk:32]: Why is this needed?

File: tests/scripts/run_tests.sh [hunk 2]
```diff
@@ -120,6 +120,7 @@ run_frequent_multi_device_pipeline_tests() {
     # Switch to modules only soon
     # run_module_tests "$tt_arch" "llrt" "$pipeline_type"
     ./tests/scripts/run_frequent_regressions_multi_device.sh
+    pytest -svv tests/tt_metal/microbenchmarks -k test_multichip_hop_latency
```
Comment (tt-rkim) [inline tests/scripts/run_tests.sh:123]: Could you please split off the specific microbenchmark into a single pytest file (such as keeping them in one place + calling only one function in a separate file)?

We're trying to redo how tests are run on the infra side, and we think we're going to go with the approach of targeting files and markers. No more `-k`

File: tests/tt_metal/tt_metal/test_kernels/dataflow/unit_tests/erisc/test_multichip_latency.cpp [hunk 3]
```diff
@@ -0,0 +1,91 @@
+// // SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
```
Comment (tt-rkim) [inline tests/tt_metal/tt_metal/test_kernels/dataflow/unit_tests/erisc/test_multichip_latency.cpp:1]: Delete this file please

File: tests/tt_metal/tt_metal/perf_microbenchmark/ethernet/test_ethernet_hop_latencies_no_edm.cpp [hunk 4]
```diff
@@ -0,0 +1,537 @@
...
+                        num_samples,
+                        sample_page_size,
+                        max_concurrent_samples,
+                        n_hops,
+
+                        programs,
+                        receiver_kernel_ids,
+                        sender_kernel_ids
+                    );
+                    // tt::tt_metal::run_roundtrip_latency_test(
```
Comment (tt-rkim) [inline tests/tt_metal/tt_metal/perf_microbenchmark/ethernet/test_ethernet_hop_latencies_no_edm.cpp:512]: Is this call needed for debug?

File: tests/tt_metal/tt_metal/perf_microbenchmark/ethernet/test_ethernet_bidirectional_bandwidth_no_edm.cpp [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+    auto [device_id, eth_receiver_core] = device_0->get_connected_ethernet_core(eth_sender_core);
+    const auto& device_1 = test_fixture.devices_.at(device_id);
+    std::cout << "sender_device_id: " << 0 << std::endl;
+    std::cout << "eth_x: " << device_0->ethernet_core_from_logical_core(eth_sender_core).x << ", " <<
+                 "eth_y: " << device_0->ethernet_core_from_logical_core(eth_sender_core).y << "" << std::endl;
+    std::cout << "sender_device_id: " << device_id << std::endl;
+    std::cout << "eth_x: " << device_1->ethernet_core_from_logical_core(eth_receiver_core).x << ", " <<
+                 "eth_y: " << device_1->ethernet_core_from_logical_core(eth_receiver_core).y << "" << std::endl;
+    // bool success = false;
+    // try {
```
Comment (tt-rkim) [inline tests/tt_metal/tt_metal/perf_microbenchmark/ethernet/test_ethernet_bidirectional_bandwidth_no_edm.cpp:315]: Are any of the commented out calls needed for debug?

File: tests/scripts/run_tests.sh (hunk 2 above)
Comment (SeanNijjar) [reply tests/scripts/run_tests.sh]: I split them up so now it's a py file for each for the latency and bandwidth benchmarks. I put them in `tests/scripts/run_frequent_regressions_multi_device.sh`  because they (well mainly the latency test) needs t3000.

File: tests/scripts/run_frequent_regressions_multi_device.sh [hunk 6]
```diff
@@ -16,6 +16,10 @@ cd $TT_METAL_HOME
 export PYTHONPATH=$TT_METAL_HOME
 
 pytest tests/ttnn/unit_tests/test_multi_device.py
+
+pytest tests/tt_metal/microbenchmarks/ethernet/test_ethernet_bidirectional_bandwidth_microbenchmark.py -k test_bidirectional_erisc_bandwidth
```
Comment (tt-rkim) [inline tests/scripts/run_frequent_regressions_multi_device.sh:20]: Hmm still looks like there's a -k here? Or are you still working on splitting it up?

Comment (SeanNijjar) [reply tests/scripts/run_frequent_regressions_multi_device.sh:20]: actually the -k is completely redundant now... I'll get rid of it :)

**Review by tt-rkim — APPROVED** [review]
